### PR TITLE
Add drag and drop to move items in FileTreeView

### PR DESCRIPTION
**Pull request by boopeshmahendran**
This PR adds the feature to drag and drop files and folders to another folder in the filetree to move them. We can drop items outside of all the folders to move items to the root folder.

I have tried to make the rename functions generic by making it accept new path instead of new name and  **reuse the rename workflow** for the move item as there is a lot common between the two things.

**Todo:-**
- [x] Add more tests
- [x] Refine the UI and fix some UI issues

@swmitra @nethip @petetnt @ficristo  @navch @shubhsnov  Please review

### CODE REVIEW

File: src/project/ProjectManager.js [hunk 1]
```diff
@@ -369,6 +377,34 @@ define(function (require, exports, module) {
...
     /**
+     * Moves the item in the oldPath to the newDirectory directory
+     *
+     * See `ProjectModel.moveItem`
+     */
+    ActionCreator.prototype.moveItem = function(oldPath, newDirectory) {
+        var self = this;
+
+        // Remove selected marker
+        self.setSelected(null);
```
Comment (boopeshmahendran) [inline src/project/ProjectManager.js]: As of now, I just remove the selected flag if any of the items are moved. Need feedback on the correct way to handle it.

**Comment by petetnt** [comment]
Really good stuff @boopeshmahendran!

Some preliminary things:

1. I think this falls under "Handle move errors", but trying to drop folder on itself brings up this:

   image
    I think it would make sense just to ignore that without showing any user errors.

2. When dragging and dropping files, it would make sense to uncollapse folders that are closed while dragging so that you can access deeply nested subfolders by just dragging. As in drag a file over a closed folder -> the folder uncollapses.

3. Not sure if you should be able to drop folders in top of the files too to move them. It's how Right click -> new folder works so it might make sense

File: src/project/ProjectManager.js [hunk 2]
```diff
@@ -369,6 +377,34 @@ define(function (require, exports, module) {
     };
 
     /**
+     * Moves the item in the oldPath to the newDirectory directory
+     *
+     * See `ProjectModel.moveItem`
+     */
+    ActionCreator.prototype.moveItem = function(oldPath, newDirectory) {
+        var self = this;
```
Comment (petetnt) [inline src/project/ProjectManager.js]: No need to set `this` to variable here (for now at least)

File: src/project/ProjectModel.js [hunk 3]
```diff
@@ -1240,6 +1254,50 @@ define(function (require, exports, module) {
...
+
+        // Add trailing slash if directory is moved
+        if (!_pathIsFile(oldPath)) {
+            fullNewPath = _ensureTrailingSlash(fullNewPath);
+        }
+
+        // Reusing the _renameItem for moving item
+        this._renameItem(oldPath, fullNewPath, fileName, !_pathIsFile(fullNewPath)).then(function () {
+            var newDirectoryRelative = self.makeProjectRelativeIfPossible(newDirectory),
+                needsLoading    = !self._viewModel.isPathLoaded(newDirectoryRelative);
```
Comment (petetnt) [inline src/project/ProjectModel.js]: Nit: some extra spaces after `needsLoading`

Comment (petetnt) [inline src/project/ProjectModel.js]: Maybe save the reference to `var viewModel = self._viewModel` to make the calls a bit less terse.

File: src/project/ProjectModel.js [hunk 4]
```diff
@@ -1240,6 +1254,50 @@ define(function (require, exports, module) {
...
+
+            // If directory view not loaded, load it and then update the view
+            if (needsLoading) {
+                self._getDirectoryContents(newDirectory).then(function(contents) {
+                    self._viewModel.setDirectoryContents(newDirectoryRelative, contents);
+                    self._viewModel.moveItem(self.makeProjectRelativeIfPossible(oldPath), self.makeProjectRelativeIfPossible(fullNewPath));
```
Comment (petetnt) [inline src/project/ProjectModel.js]: Maybe make these variables too `self.makeProjectRelativeIfPossible(oldPath)` `self.makeProjectRelativeIfPossible(fullNewPath)` as the long lines are pretty hard to distinquish.

**Review by petetnt — COMMENTED** [review]
Some really small comments, really nice job 👍

**Comment by boopeshmahendran** [comment]
@petetnt Thanks for reviewing. I'll address these comments.

**Comment by boopeshmahendran** [comment]
@petetnt I have addressed your comments. Please review.

File: src/filesystem/FileIndex.js [hunk 5]
```diff
@@ -138,6 +142,30 @@ define(function (require, exports, module) {
...
+        // If file path is changed, i.e the file is moved
+        // Remove the moved entry from old Directory and add it to new Directory
+        if (oldParentPath !== newParentPath) {
+            var oldDirectory = this._index[oldParentPath],
+                newDirectory = this._index[newParentPath],
+                renamedEntry;
+
+            if (oldDirectory && oldDirectory._contents) {
+                oldDirectory._contents = oldDirectory._contents.filter(function(entry) {
+                    if (entry.fullPath === oldPath || entry.fullPath === newPath) {
```
[the remark this answers is not in the document]
Comment (boopeshmahendran) [reply src/filesystem/FileIndex.js]: Yeah. Changed this.

**Review by petetnt — COMMENTED** [review]
Codewise LGTM, didn't have the time to re-test the functionality yet

**Review by nethip — APPROVED** [review]

**Pull request merged**